chain/db/statedb.go

```diff
@@ -2,14 +2,279 @@ package db
 
 import (
 	"bytes"
+	"crypto/sha256"
 	"fmt"
 	"io"
+	"sort"
 	"sync"
 
 	"github.com/nknorg/nkn/common"
 	"github.com/nknorg/nkn/common/serialization"
+	"github.com/nknorg/nkn/util"
+	"github.com/nknorg/nkn/util/config"
 )
 
+type nanoPay struct {
+	balance common.Fixed64
+	limit   uint32
+}
+
+func (np *nanoPay) Serialize(w io.Writer) error {
+	err := np.balance.Serialize(w)
+	if err != nil {
+		return fmt.Errorf("account nanoPay Serialize error: %v", err)
+	}
+
+	err = serialization.WriteUint32(w, np.limit)
+	if err != nil {
+		return fmt.Errorf("account nanoPay Serialize error: %v", err)
+	}
+
+	return nil
+}
+
+func (np *nanoPay) Deserialize(r io.Reader) error {
+	var balance common.Fixed64
+	err := balance.Deserialize(r)
+	if err != nil {
+		return fmt.Errorf("Deserialize nanoPay error: %v", err)
+	}
+	np.balance = balance
+
+	np.limit, err = serialization.ReadUint32(r)
+	if err != nil {
+		return fmt.Errorf("Deserialize nanoPay error: %v", err)
+	}
+
+	return nil
+}
+
+var globalAddr, _ = common.ToScriptHash(config.GlobalAccountAddress)
+
+var newGlobalAccount = func() serialization.SerializableData {
+	return new(globalAccount)
+}
+
+type globalAccount struct {
+	nanoPay        map[string]*nanoPay
+	nanoPayCleanup map[uint32]map[string]struct{}
+}
+
+func NewGlobalAccount() *globalAccount {
+	return &globalAccount{nanoPay: make(map[string]*nanoPay), nanoPayCleanup: make(map[uint32]map[string]struct{})}
+}
+
+func (ga *globalAccount) Serialize(w io.Writer) error {
+	var nps []string
+	for k := range ga.nanoPay {
+		nps = append(nps, k)
+	}
+	sort.Strings(nps)
+
+	if err := serialization.WriteVarUint(w, uint64(len(nps))); err != nil {
+		return fmt.Errorf("account nanoPay Serialize error: %v", err)
+	}
+
+	for _, k := range nps {
+		v := ga.nanoPay[k]
+
+		err := serialization.WriteVarString(w, k)
+		if err != nil {
+			return fmt.Errorf("globalAccount nanoPay Serialize error: %v", err)
+		}
+
+		err = v.Serialize(w)
+		if err != nil {
+			return fmt.Errorf("globalAccount nanoPay Serialize error: %v", err)
+		}
+	}
+
+	var npcs util.Uint32Slice
+	for k := range ga.nanoPayCleanup {
+		npcs = append(npcs, k)
+	}
+	sort.Sort(npcs)
+
+	if err := serialization.WriteVarUint(w, uint64(len(npcs))); err != nil {
+		return fmt.Errorf("globalAccount nanoPayCleanup Serialize error: %v", err)
+	}
+
+	for _, k := range npcs {
+		if err := serialization.WriteUint32(w, k); err != nil {
+			return fmt.Errorf("globalAccount nanoPayCleanup Serialize error: %v", err)
+		}
+
+		npc := ga.nanoPayCleanup[k]
+		if err := serialization.WriteVarUint(w, uint64(len(npc))); err != nil {
+			return fmt.Errorf("globalAccount nanoPayCleanup Serialize error: %v", err)
+		}
+
+		for id := range npc {
```

Comment:
this npc is type of 'map[string]struct{}', the keys of it should be out-of-order.

Reply:
I think the plan is to change it to trie from map so we don't need this anymore

Reply:
@yilunzhang is right, we won't have that serialization anymore